File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py

def dfs(start,check=[]):
    print(start,end=' ')
    check.append(start)
    for i in mat[start]:
        if i not in check:
            dfs(i)
            check.append(i)
    
    
    
def bfs(start):
    print()
    check=[start]
    queue=[start]
    tum,popi=0,1
    while len(queue)>0:
        popi=queue[0]
        queue=queue[1:]
        print(popi,end=" ")
        for i in mat[popi]:
            if i not in check:
                queue.append(i)
                check.append(i)
        

N,M,V=map(int,input().split())
mat=[[] for _ in range(N+1)]
for _ in range(M):
    a,b=map(int,input().split())
    mat[a].append(b)
    mat[b].append(a)
for i in range(len(mat)):
    mat[i].sort()
dfs(V)
bfs(V)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.cc
#include <iostream>
#include <vector>
#include <utility>
#include <algorithm>
#include <string.h>
#include <queue>
using namespace std;

int N, M, V;
vector<pair<int, int> > a;
bool visited[1001];

bool compare(pair<int, int> a, pair<int, int> b) {
	if (a.first == b.first) return a.second < b.second;
	else return a.first < b.first;
}


void dfs(int t) {
	cout << t << " ";
	visited[t] = true;
	for (int i = 0; i < M; i++) {
		if (a[i].first == t && visited[a[i].second] == false) {
			dfs(a[i].second);
		}
		else if (a[i].second == t && visited[a[i].first] == false) {
			dfs(a[i].first);
		}
	}
}

void bfs(int t) {
	queue<int> q;
	q.push(t);
	visited[t] = true;
	int x;

	while (q.empty() == false) {
		x = q.front();
		cout << x << " ";
		q.pop();
		for (int i = 0; i < M; i++) {
			if (a[i].first == x && visited[a[i].second] == false) {
				q.push(a[i].second);
				visited[a[i].second] = true;
			}
			else if (a[i].second == x && visited[a[i].first] == false) {
				q.push(a[i].first);
				visited[a[i].first] = true;
			}
		}

	}

}


int main() {

	cin >> N >> M >> V;

	int x, y;

	memset(visited, false, sizeof(visited));

	for (int i = 0; i < M; i++) {
		cin >> x >> y;
		if (x == y) continue;
		else if (x > y) a.push_back(make_pair(y, x));
		else a.push_back(make_pair(x, y));
	}
	sort(a.begin(), a.end(), compare);
		dfs(V);
		cout << endl;
		memset(visited, false, sizeof(visited));
		bfs(V);

	return 0;
}